Add support for ESLint 9.x

jeddy3 opened the pull request:
<!-- Each pull request must be associated with an open issue unless it's a documentation fix. If a corresponding issue does not exist, please create one so we can discuss the change first. -->

<!-- Please answer the following. We close pull requests that don't. -->

> Which issue, if any, is this issue related to?

Closes #297

> Is there anything in the PR that needs further explanation?

I figured I'd use this as an opportunity to familiarise myself more with ESLint's flat config.

- [9.x Migration Guide](https://eslint.org/docs/latest/use/migrate-to-9.0.0)
- [Configuration Migration Guide](https://eslint.org/docs/latest/use/configure/migration-guide)

ybiquitous commented on jest.js:
```diff
@@ -1,5 +1,5 @@
-};
+const jest = require('eslint-plugin-jest');
+
+module.exports = [{ ...jest.configs['flat/recommended'] }, { ...jest.configs['flat/style'] }];
```
[suggestion] Minor. When seeing <https://github.com/stylelint/eslint-config-stylelint/pull/298/files?diff=split&w=1#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R10-R12>, it seems to be able simplify this file:

```suggestion
module.exports = [jest.configs['flat/recommended'], jest.configs['flat/style']];
];
```

ybiquitous reviewed (commented):
Thank you, overall looks great!

Can you take a look at my comments?

jeddy3 commented:
@ybiquitous Thank you for the review. I've made the three changes you suggested. I've also:

- updated the README
- added changelog entries

So that we're ready to release.

ybiquitous commented on eslint.config.js:
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+const config = require('./index.js');
+
+module.exports = [
+	...config,
```
[suggestion] I'm wondering if testing `jest.js` in the original `.eslintrc.js` might still be valuable. What do you think?

```suggestion
const config = require('./index.js');
const jestConfig = require('./jest.js');

module.exports = [
	...config,
	...jestConfig,
```

jeddy3 replied:
We use `eslint.config.js` to configure ESLint for the repo, i.e. to lint the configs and their tests. The tests use `node:test` rather than Jest. Perhaps we accidentally left in the jest config in the original `.eslintrc` when we converted the tests to use `node:test`?

ybiquitous replied:
Ah, got it. We've already tested `jest.js` in `__tests__/jest.test.js`, so my suggestion is unnecessary. 👍🏼

ybiquitous reviewed (commented):
Thanks for the quick action! 👍🏼 
In addition, I noticed two possible improvements. Can you see them?

jeddy3 commented:
Thank you for the further review. I totally forgot about exports 😅 !

ybiquitous reviewed (approved):
Thanks, LGTM 👍🏼

Mouvedia commented on package.json:
Stylelint uses `>=18.12`, it won't matter?

ybiquitous replied:
Anyway, we must respect ESLint's `engines.node`.

ESLint is a dev dependency, so I think it's not a problem. Here's the current dep tree of Stylelint:

```sh-session
$ npm ls espree
stylelint@16.10.0 /Users/masafumi/git/stylelint/stylelint
├─┬ eslint-config-stylelint@22.0.0
│ └─┬ @stylistic/eslint-plugin-js@2.3.0
│   └── espree@10.1.0
└─┬ eslint@8.57.0
  ├─┬ @eslint/eslintrc@2.1.4
  │ └── espree@9.6.1 deduped
  └── espree@9.6.1

$ npm v espree@10.1.0 engines.node
^18.18.0 || ^20.9.0 || >=21.1.0
```
